openedx/openedx-platform
# [BB-1847] Simple theme improvements and customizations

This PR adds fixes and additional SASS variables to the default Open edX theme in order to apply styles uniformly to bootstrap and non-bootstrap pages and also easily customize the colors and styles. This includes the changes that were demonstrated during the October 2019 community hangouts.

**JIRA tickets**: None

**Discussions**: https://discuss.openedx.org/t/open-edx-theming-improvements-and-simplification/461

**Dependencies**: None

**Sandbox URL**: TBD - sandbox is being provisioned.

**Merge deadline**: None

**Testing instructions**:
1. Clone the source branch of this PR.
1. Compile the default theme and verify that it looks and behaves similar to the current vanilla theme.
2. Customize the theming by providing values for the SASS variables for the newly added variables and verify that they work as expected and are applied uniformly.  [This skeleton theme](https://github.com/open-craft/edx-simple-theme/tree/guruprasad/BB-1847-new-simple-theme-skeleton) with the variables added in `lms/static/sass/common-variables.scss` can be used for testing. An example `common-variables.scss` file is [here](https://gist.github.com/giovannicimolin/96d5859673b0dd4b70fd6f95ee652d3f).

**Author notes, concerns and questions**:
1. The biggest challenge in this work is to retain the existing defaults while providing a way to easily customize the colors and styles. In many cases, the defaults are defined in external libraries (edX patter library for example) or using mixins, inheritance, hard-coded SASS variables and are available in only one of the two theming systems. So duplicating the default definitions is often required to make the styles apply uniformly across the systems. A way to avoid this is to use SASS conditions like `@if variable-exists(my-custom-variable)` to conditionally use the new SASS variables. But that will become unmaintainable and ugly when used for a lot of variables.
1. A lot of the SASS variables added in this PR, target specific elements of various pages and hence are tightly coupled to the current UI implementation. When more components get migrated to React microfrontends, which is likely to have its own theming, the UI changes will break the changes made here. Even if it is possible to fix the issues, 3 theming systems will be in use till the whole codebase is updated to use a single theming system and that is likely going to take a lot of time.
1. Is there a better way to do this? These changes were initially created for a theme skeleton to be used along with the `simple-theme` role in the configuration repository. However, we want to upstream the changes so that everyone benefits from the changes and also to reduce the maintenance work required to keep the theme skeleton working for every new release of Open edX.

    One option that we considered was to propose including this as one of the default themes that ship in the edx-platform codebase, however, that would also mean that the improvements made for this will not be usable by most users of the platform unless they specifically switch to this theme.
1. The default theme defines a color palette which has colors like `primary`, `inverse`, `dark`, `light` etc. defined. However, these are neither used everywhere consistently nor there are guidelines for where to use each of those. In some places, some SASS color variables are used and there are comments which indicate that they are used to meet the WCAG guidelines. But these are not used everywhere and hence it is not clear if those can and should be overridden.

**Known issues and limitations**:
* This PR doesn't attempt to add customizations for every page in the LMS or every element in all the pages it customizes. It only tries to allow get an uniform application of colors. The details of what pages can be themed and what cannot be is mentioned in [this review comment](https://github.com/edx/edx-platform/pull/22543#pullrequestreview-362180360).
* If possible, I will try to add similar customizations to a few more low-hanging fruits.
* The `stylelint` checks are failing in the files modified by this PR but all the errors corresponding to the modified lines have been fixed. But the linter still lists the errors in other unmodified lines in those files.
* If in the future, someone else changes the pages customized by this PR but does not use allow customization for any new changes, those will look odd and out of place. So the style of changes made by this PR will not be very useful if it is a one-off change and all the future changes don't care about the customizability.

**Reviewers**
